src/IO/BoundedReadBuffer.cpp

```diff
@@ -4,8 +4,7 @@
 namespace DB
 {
 
-BoundedReadBuffer::BoundedReadBuffer(std::unique_ptr<SeekableReadBuffer> impl_)
-    : ReadBufferFromFileDecorator(std::move(impl_))
+BoundedReadBuffer::BoundedReadBuffer(std::unique_ptr<ReadBufferFromFileBase> impl_) : impl(std::move(impl_))
```

Comment:
It looks there is an issue, since `swap` willl be calleed in `ReadBufferFromFileDecorator`'s constructor. 

Without `swap`,  the following **assert** will fail.
```c++
bool ReadBufferFromFileDescriptor::nextImpl()
{
    /// If internal_buffer size is empty, then read() cannot be distinguished from EOF
    assert(!internal_buffer.empty());
    ...
}
```

Reply:
Could you provide a reproducible example?

Reply:
> Could you provide a reproducible example?

In ClickHouse, `BoundedReadBuffer` is only used in `CachedOnDiskReadBufferFromFile`, I don't know how to reproduce it. However, let's say we want to read 100 to 1000 bytes of a file

``` c++
auto read_buffer = std::make_unique<DB::ReadBufferFromFilePRead>("xxx");
read_buffer = std::make_unique<DB::BoundedReadBuffer>(std::move(read_buffer));
read_buffer->seek(100, SEEK_SET);
read_buffer->setReadUntilPosition(1000);
```
The above code is definitely wrong.

Reply:
@CurtizJ any updates？